packages/runtime-core/src/componentOptions.ts

```diff
@@ -506,9 +507,7 @@ export function applyOptions(
     errorCaptured,
     // public API
     expose
-  } = options
-
-  const publicThis = instance.proxy!
+  } = asMixin ? options : publicThis.$options
```

Comment:
The `$options` is a merged options, and the merge strategy is decided by the users, this means:
1. Unreliable
Users can return any value by customizing the merge strategy, we need to check them.
2. Logic conflict with applyOptions
Now, `applyOptions` is called recursively when initializing options, but `$options` may contain options from mixins/extends, do we still need recursive calls?

I think we need to carefully redesign, but indeed, this PR looks more reasonable, compared to https://github.com/vuejs/vue-next/pull/2792